Add module table builder

sormuras opened the pull request:
Run with: `jshell build.jsh`

[generated/modules.md](https://github.com/sormuras/jpms-module-names/blob/builder/generated/modules.md)

- [x] Add unit tests
- [x] Add already listed projects
- [x] Introduce `pom.xml` parser as default source for project information
- [x] Generate CSV, TSV and `module-maven.properties` + [`module-version.properties`](https://sormuras.github.io/blog/2017-12-13-module-info-properties.html) files

sormuras committed:
Add module table builder

Run with: `jshell build.jsh`

jodastephen commented on .gitignore:
```diff
@@ -1,3 +1,8 @@
-csv/.gradle/
-csv/build/
-*.class
+# Gradle
```
This needs merging with the existing gitignore

sormuras replied:
Logically it supersedes the existing one. Or did I miss some files?

jodastephen replied:
Is the csv folder gone? I think its still there...

sormuras replied:
Entries in `.gitignore` that don't start with a `/` are applied to every sub-folder.

jodastephen commented on builder/Builder.java:
```diff
@@ -0,0 +1,382 @@
+    builder
+        .add("org.joda", "joda-beans")
+        .project("Joda-Beans")
+        .homepage("http://www.joda.org/joda-beans");
```
Maybe the URL could be extracted from the pom.xml

sormuras replied:
Good idea - as long there is a) an associated `pom.xml` and b) the URI is specified there.

sormuras replied:
I'll add and use that pom-parsing feature when I implement a logical table section builder.

Something like this:

```java
var section = Section.of("org.joda").project("Joda").homepage("..."). // Section.of(<pom.xml>).
builder.add(section)
```

jodastephen commented on builder/Builder.java:
```diff
@@ -0,0 +1,382 @@
+                .map(StandardCharsets.UTF_8::decode)
+                .map(Object::toString)
+                .orElse("");
+        if (manifestString.contains("Automatic-Module-Name")) {
```
Will need to check for `module-info.class` too, unless I missed it.

sormuras replied:
That check is _a priori_ done by the `ModuleFinder` usage. Only "automatic" modules are required to be stable, i.e. define an `Automatic-Module-Name` attribute.

sormuras replied:
I'll add a flag to the table that indicates whether a module is explicit or autotmatic.

sormuras commented:
Btw. is it okay if I switch to Java 10 `var` syntax?

jodastephen commented:
Are we going to need it to run on travis? Anything beyond Java 9 not
supported yet.

On 21 Dec 2017 3:30 p.m., "Christian Stein" <notifications@github.com>
wrote:

> Btw. is it okay if I switch to Java 10 var syntax?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jodastephen/jpms-module-names/pull/17#issuecomment-353379626>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANA3B3p85f87tnsf6xLK75JqB_gJdPtks5tCnl8gaJpZM4RJygE>
> .
>

sormuras commented:
Using [install-jdk-10](https://sormuras.github.io/blog/2017-12-08-install-jdk-on-travis.html) it is easy to setup -- but before running on a CI environment, we need a "don't look up already mapped artifacts" feature. Otherwise `mvnrepository` auto-blocks that IP for ... how long I don't know. ;)

sormuras committed:
Flag modules as automatic or explicit

sormuras commented:
Current table versios can viewed here: https://github.com/sormuras/jpms-module-names/blob/builder/generated/modules.md

sormuras committed:
Add ASM Core 6.0 as an example for an explicit module

sormuras committed:
Add more output formats

* Comma Separated Values:  generated/modules.csv
* Tab Separated Values: generated/modules.tsv
* Properties module name to Maven coordinates: module-maven.properties
* Properties module name to released versions: module-version.properties

sormuras committed:
Rename Builder.Line to Builder.Item

sormuras committed:
Add unit tests and add POM parsing and use it where applicable

sormuras committed:
Add remaining well-known projects to the build script

sormuras commented:
All generated formats can be viewed here: https://github.com/sormuras/jpms-module-names/tree/builder/generated

- Group crawling, logical section support and Java 10 syntax are not happening soon. ;)
- The `csv` directory can now be marked "deprecated" or removed entirely.
- `BuilderTests` are runnable from within IDEA after marking "builder/" as a source directory and adding "JUnit 5" to the referenced libraries -- a "headless" test task will follow

Anything left to do?

sormuras committed:
Add more ASM and Netty modules

sormuras committed:
Add more Spring modules

sormuras committed:
Cleaner Markdown layout

- Promote explicit or automatic module icon as emoji
- Use fixed-width font type for module names
- Join Maven Group ID, Artifact ID and Version cells
